core/embed/io/ble/stm32/ble.c

```diff
@@ -0,0 +1,486 @@
+    return;
+  }
+
+  event_status_msg_t msg;
+  memcpy(&msg, data, sizeof(event_status_msg_t));
+
+  if (!drv->status_valid) {
+    if (msg.peer_count > 0) {
+      drv->mode_requested = BLE_MODE_CONNECTABLE;
+    } else {
+      drv->mode_requested = BLE_MODE_OFF;
+    }
+  }
+
+  if (drv->connected != msg.connected) {
+    if (msg.connected) {
+      // new connection
+
+      ble_event_t event = {.type = BLE_CONNECTED};
+      tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event),
+                      NULL);
+
+    } else {
+      // connection lost
+      ble_event_t event = {.type = BLE_DISCONNECTED};
+      tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event),
+                      NULL);
+
+      if (drv->mode_current == BLE_MODE_PAIRING) {
+        drv->mode_requested = BLE_MODE_CONNECTABLE;
+      }
+    }
+
+    drv->connected = msg.connected;
+  }
+
+  if (msg.advertising && !msg.advertising_whitelist) {
+    drv->mode_current = BLE_MODE_PAIRING;
+  } else if (msg.advertising) {
+    drv->mode_current = BLE_MODE_CONNECTABLE;
+  } else {
+    drv->mode_current = BLE_MODE_OFF;
+  }
+
+  drv->peer_count = msg.peer_count;
+
+  drv->status_valid = true;
+}
+
+static void ble_process_rx_msg_pairing_request(const uint8_t *data,
+                                               uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+  if (!drv->initialized) {
+    return;
+  }
+
+  if (len < 7) {
+    // insufficient data length
+    return;
+  }
+
+  if (drv->mode_requested != BLE_MODE_PAIRING ||
+      drv->mode_current != BLE_MODE_PAIRING) {
+    ble_send_pairing_reject();
+    return;
+  }
+
+  ble_event_t event = {.type = BLE_PAIRING_REQUEST, .data_len = 6};
+  memcpy(event.data, &data[1], 6);
+  tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event), NULL);
+}
+
+static void ble_process_rx_msg_pairing_cancelled(const uint8_t *data,
+                                                 uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+  if (!drv->initialized) {
+    return;
+  }
+
+  ble_event_t event = {.type = BLE_PAIRING_CANCELLED, .data_len = 0};
+  tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event), NULL);
+}
+
+static void ble_process_rx_msg(const uint8_t *data, uint32_t len) {
+  if (len < 1) {
+    return;
+  }
+
+  switch (data[0]) {
+    case INTERNAL_EVENT_STATUS:
+      ble_process_rx_msg_status(data, len);
+      break;
+    case INTERNAL_EVENT_PAIRING_REQUEST:
+      ble_process_rx_msg_pairing_request(data, len);
+      break;
+    case INTERNAL_EVENT_PAIRING_CANCELLED:
+      ble_process_rx_msg_pairing_cancelled(data, len);
+      break;
+    default:
+      break;
+  }
+}
+
+static void ble_process_data(const uint8_t *data, uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+
+  if (len != BLE_RX_PACKET_SIZE) {
+    return;
+  }
+
+  tsqueue_enqueue(&drv->rx_queue, data, len, NULL);
```

Comment:
I'm a bit concerned about silently dropping data when a queue is full. Here it's probably OK, we haven't sent application-level ACK yet so Host will time out and resend the packet as if it got lost in the air.

`tx_queue` overflows get reported to the upper layer so that's OK too.

Not sure about `event_queue` - if events get dropped UI will show incorrect BLE state. This could be avoided by UI by ignoring the type of the event and always asking the driver for the entire BLE state, but I think we had this with USB and changed it?

Wondering if we should maintain an overflow counter for every queue in debug builds, along with a function to retrieve them.